oak_attestation_verification/src/util.rs

```diff
@@ -94,9 +96,123 @@ pub fn verify_signature_raw(
 }
 
-/// Computes a SHA2-256 digest of `input` and returns it as raw bytes.
 pub fn hash_sha2_256(input: &[u8]) -> [u8; 32] {
     let mut hasher = Sha256::new();
     hasher.update(input);
     hasher.finalize().into()
 }
+
+fn hash_sha2_512(input: &[u8]) -> [u8; 64] {
+    let mut hasher = Sha512::new();
+    hasher.update(input);
+    hasher.finalize().into()
+}
+
+fn hash_sha2_384(input: &[u8]) -> [u8; 48] {
+    let mut hasher = Sha384::new();
+    hasher.update(input);
+    hasher.finalize().into()
+}
+
+/// Computes various digest formats of a binary array.
+/// The empty arrays need to be filled, when we depend on SHA{1,3} hashers.
+pub fn raw_digest_from_contents(contents: &[u8]) -> RawDigest {
+    RawDigest {
+        psha2: b"".to_vec(),
+        sha1: b"".to_vec(),
+        sha2_256: hash_sha2_256(contents).to_vec(),
+        sha2_512: hash_sha2_512(contents).to_vec(),
+        sha3_512: b"".to_vec(),
+        sha3_384: b"".to_vec(),
+        sha3_256: b"".to_vec(),
+        sha3_224: b"".to_vec(),
+        sha2_384: hash_sha2_384(contents).to_vec(),
+    }
+}
+
+#[derive(PartialEq)]
+pub enum MatchResult {
+    SAME = 0,
+    DIFFERENT = 1,
+    UNDECIDABLE = 2,
+}
+
+/// Compares two binary arrays byte per byte.
+fn match_strings(a: &str, b: &str) -> MatchResult {
+    if a != b {
+        return MatchResult::DIFFERENT;
+    }
+    return MatchResult::SAME;
+}
+
+/// Compares two hex digests.
+pub fn is_hex_digest_match(a: &HexDigest, b: &HexDigest) -> MatchResult {
+    if !a.psha2.is_empty() && !b.psha2.is_empty() {
+        return match_strings(&a.psha2, &b.psha2);
+    }
+    if !a.sha2_256.is_empty() && !b.sha2_256.is_empty() {
+        return match_strings(&a.sha2_256, &b.sha2_256);
+    }
+    if !a.sha2_512.is_empty() && !b.sha2_512.is_empty() {
+        return match_strings(&a.sha2_512, &b.sha2_512);
+    }
+    if !a.sha3_512.is_empty() && !b.sha3_512.is_empty() {
+        return match_strings(&a.sha3_512, &b.sha3_512);
+    }
+    if !a.sha3_384.is_empty() && !b.sha3_384.is_empty() {
+        return match_strings(&a.sha3_384, &b.sha3_384);
+    }
+    if !a.sha3_256.is_empty() && !b.sha3_256.is_empty() {
+        return match_strings(&a.sha3_256, &b.sha3_256);
+    }
+    if !a.sha3_224.is_empty() && !b.sha3_224.is_empty() {
+        return match_strings(&a.sha3_224, &b.sha3_224);
+    }
+    if !a.sha2_384.is_empty() && !b.sha2_384.is_empty() {
+        return match_strings(&a.sha2_384, &b.sha2_384);
+    }
+
+    // Nit: Put the weak hash to the end of comparisons.
+    if !a.sha1.is_empty() && !b.sha1.is_empty() {
+        return match_strings(&a.sha1, &b.sha1);
+    }
+
+    return MatchResult::UNDECIDABLE;
+}
+
+/// Compares two raw digests.
+pub fn is_raw_digest_match(a: &RawDigest, b: &RawDigest) -> MatchResult {
+    return is_hex_digest_match(&raw_to_hex_digest(a), &raw_to_hex_digest(b));
+}
+
+/// Converts raw digest to hex digest.
+pub fn raw_to_hex_digest(r: &RawDigest) -> HexDigest {
```

Comment:
Personally I would have done the conversion the other way, i.e. from hex directly to binary (raw), so that if it does not parse correctly you also have more information early on. The hex representation is only for human readability, IMO it should not be used internally, especially not for comparisons.

Reply:
The hex representation is used internally, whenever a digest occurs in a JSON file (which is frequently).